akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -143,7 +146,14 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
     def shutdown(): IncomingStreamState
   }
   case object Idle extends IncomingStreamState {
-    def handle(event: StreamFrameEvent): IncomingStreamState = expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
+    def handle(event: StreamFrameEvent): IncomingStreamState = {
+      if (incomingStreams.size > settings.maxConcurrentStreams) {
+        resetStream(event.streamId, ErrorCode.REFUSED_STREAM)
+        Closed
+      } else {
+        expectIncomingStream(event, HalfClosedRemote, ReceivingDataFirst)
+      }
+    }
```

Comment:
this is the change where the setting is enforced. I only implemented enforcing on `Idle` instead of adding it into `expectIncomingStream` because `expectIncomingStream` is invoked from other states. This solution makes it possible to decrease the value of `maxConcurrentStreams` (in a future PR) and let the in-flight streams to finish.

Reply:
But that assumes that every received event here will open a stream. At least `PRIORITY` frames can be received in this stream without opening it.

> this is the change where the setting is enforced. I only implemented enforcing on `Idle` instead of adding it into `expectIncomingStream` because `expectIncomingStream` is invoked from other states. This solution makes it possible to decrease the value of `maxConcurrentStreams` (in a future PR) and let the in-flight streams to finish.

You could also state it differently: `maxConcurrentStreams` is directional and should only account for streams opened from the peer. Right now we count all streams which is only correct for the server side in cases without `PUSH_PROMISE`. On the client side, we move out of this state using `handleOutgoingCreated` so we side-step the issue.

Maybe it's enough to add a check that the event actually is a `ParsedHeaderFrame`?

Reply:
I didn't consider `PRIORITY` frames and I agree they should not trigger the verification. After a quick review, it makes sense to me to only check the number of concurrent streams when processing a `ParsedHeaderFrame`.

We may need to revisit this logic when supporting push but we'll cross that bridge when we get there.